From: Hannes Reinecke <hare@suse.de>
Date: Tue, 25 Feb 2014 08:04:42 +0100
Subject: scsi_dh_alua: endless STPG retries for a failed LUN
Patch-Mainline: Not yet
References: bnc#865342

When a LUN fails RTPG will fail with an error, but this will
be overwritten later on when STPG is called. This might
trigger an endless cycle.
Instead we should be evaluating any errors from RTPG, and
do not call STPG if RTPG fails.

Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 drivers/scsi/device_handler/scsi_dh_alua.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/scsi/device_handler/scsi_dh_alua.c b/drivers/scsi/device_handler/scsi_dh_alua.c
index 16877f8..017d0ee 100644
--- a/drivers/scsi/device_handler/scsi_dh_alua.c
+++ b/drivers/scsi/device_handler/scsi_dh_alua.c
@@ -957,6 +957,8 @@ static void alua_rtpg_work(struct work_struct *work)
 		}
 		spin_lock_irqsave(&pg->rtpg_lock, flags);
 		pg->flags &= ~ALUA_PG_RUN_RTPG;
+		if (err != SCSI_DH_OK)
+			pg->flags &= ~ALUA_PG_RUN_STPG;
 	}
 	if (pg->flags & ALUA_PG_RUN_STPG) {
 		spin_unlock_irqrestore(&pg->rtpg_lock, flags);
-- 
1.7.12.4

